### PR TITLE
Scoring cities

### LINKED ISSUE #19: Implement scoring cities
2 functions for scoring the features:
- score completed feature after a move is made (the tile that was just placed is passed to the function)
- score uncompleted features after the game finishes

https://wikicarpedia.com/car/Base_game#3._Scoring_a_city

### PR DESCRIPTION
**Pull request by BognaLew**
Closes #19

### CODE REVIEW

File: pkg/game/city/city.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+	// award points
+	for _, player := range winningPlayers {
+		scoreReport.ReceivedPoints[player] = totalScore
+	}
+
+	return scoreReport
+}
+
+// Returnes all features from a tile at a given position that are part of a city
```
Comment (bstrzelecki) [inline pkg/game/city/city.go:127]:
```suggestion
// Returns all features from a tile at a given position that are part of a city
```

File: pkg/game/city/city.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	return cities, ok
+}
+
+func (city *City) AddTile(pos elements.Position, cityFeatures []elements.PlacedFeature, hasShield bool) {
+	city.cities[pos] = cityFeatures
+	city.shields[pos] = hasShield
+	city.checkCompleted()
+}
+
+// Merges two cities when they are connetced.
```
Comment (bstrzelecki) [inline pkg/game/city/city.go:140]:
```suggestion
// Merges two cities when they are connected.
```

File: pkg/game/city/city_manager.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+				}
+			}
+			mask = mask.Rotate(1)
+		}
+		citiesToJoin[cityFeature] = cToJoin
+	}
+	return citiesToJoin
+}
+
+// Performes required operations to add a new city feature.
```
Comment (bstrzelecki) [inline pkg/game/city/city_manager.go:79]:
```suggestion
// Performs required operations to add a new city feature.
```

File: pkg/game/city/city.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		cities: map[elements.Position][]elements.PlacedFeature{
+			pos: cityFeature,
+		},
+		shields: map[elements.Position]bool{
+			pos: hasShield,
+		},
+	}
+}
+
+func (city City) GetCompleted() bool {
```
Comment (bstrzelecki) [inline pkg/game/city/city.go:27]: Maybe IsCompleted?

File: pkg/game/city/city.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+func (city *City) checkCompleted() bool {
+	city.completed = true
+	for key, placedFeatures := range city.cities {
+		for _, placedFeature := range placedFeatures {
+			sides := placedFeature.Feature.Sides
+			mask := side.Top
+			for range 4 {
+				if sides&mask == mask {
+					switch mask {
+					case side.Top:
```
Comment (bstrzelecki) [inline pkg/game/city/city.go:41]: Using [PositionFromSide](https://github.com/YetAnotherSpieskowcy/Carcassonne-Engine/blob/f9fd846edf803179db7d10676847d1395887284c/pkg/game/elements/placed_tile.go#L39) would by slightly cleaner.

**Review by bstrzelecki — DISMISSED** [review]

**Review by bstrzelecki — DISMISSED** [review]

File: pkg/game/city/city_manager_test.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	"github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tiles/tiletemplates"
+)
+
+func TestGetNeighbouringPositions(t *testing.T) {
+	positions := getNeighbouringPositions(elements.NewPosition(1, 1))
+
+	topPosition := positions[side.Top]
+	if topPosition.X() != 1 || topPosition.Y() != 2 {
+		t.Fatalf("expected x=%#v y=%#v, got x=%#v y=%#v instead", 1, 2, topPosition.X(), topPosition.Y())
+	}
```
Comment (Lenetis) [inline pkg/game/city/city_manager_test.go:18]: This should probably check all positions instead of just top

**Review by Lenetis — CHANGES_REQUESTED** [review]

**Comment by Lenetis** [comment]
Also, consider adding a test for this kind of city setup:
image

File: pkg/game/city/city_manager.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+func (manager *Manager) UpdateCities(tile elements.PlacedTile) {
+	foundCities := manager.findCities(tile.Position)
+
+	if len(foundCities) > 0 {
+		citiesToRemove := make([]int, 0)
+		citiesToJoin := manager.findCitiesToJoin(foundCities, tile)
+		// join cities
+		for f, cToJoin := range citiesToJoin {
+			if len(cToJoin) == 0 {
+				// I guess panic
```
Comment (Lenetis) [inline pkg/game/city/city_manager.go:89]: Yup, I think panic would be alright here

File: pkg/game/city/city_manager.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+				break
+			}
+			if len(cToJoin) > 1 {
+				for _, cityIndex := range cToJoin[1:] {
+					manager.cities[cToJoin[0]].JoinCities(manager.cities[cityIndex])
+					citiesToRemove = append(citiesToRemove, cityIndex)
+					println("aaa")
```
Comment (Lenetis) [inline pkg/game/city/city_manager.go:96]: Probably to remove :)

File: pkg/game/city/city_manager.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+	} else {
+		for _, f := range tile.GetCityFeatures() {
+			toAppend := []elements.PlacedFeature{f}
+			manager.cities = append(manager.cities, NewCity(tile.Position, toAppend, tile.HasShield))
+		}
+	}
+}
+
+// Calculates ScoreReport. When forceScore = false calculates score only based on
+// closed cities and sets city.score to true. Otherwise calculates score based on
```
Comment (Lenetis) [inline pkg/game/city/city_manager.go:121]:
```suggestion
// closed cities and sets city.scored to true. Otherwise calculates score based on
```

**Review by Lenetis — APPROVED** [review]

**Review by bstrzelecki — APPROVED** [review]

**Pull request merged**